File: 39. Combination Sum/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<List<Integer>> res = new ArrayList<>();
        getComb(res,new ArrayList<>(),target,0,candidates);
        return res;
    }
    
    private void getComb(List<List<Integer>> res, List<Integer> current, int target, int pos, int[] candidates ){
        System.out.println(target);
        if(target == 0) res.add(new ArrayList<>(current));   //copy arrayList
        else if(target < 0) return;
        else{
            for(int i = pos; i < candidates.length; i++){
                current.add(candidates[i]);
                getComb(res,current,target - candidates[i],i,candidates);
                current.remove(current.size() - 1);
            }
            
        }
    }
    
}

File: js/array/39. Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 * backtracking
 */
var combinationSum = function(candidates, target) {
  let res = [];
  let current = [];
  getComb(res, current, candidates, target, 0);
  return res;
};

const getComb = function(res, current, nums, target, pos) {
  if (target < 0) return;
  else if (target == 0) res.push(current.slice());
  //copy array
  else {
    for (let i = pos; i < nums.length; i++) {
      current.push(nums[i]);
      getComb(res, current, nums, target - nums[i], i);
      current.pop();
    }
  }
};
